modules.go

```diff
@@ -0,0 +1,455 @@
+package goja
+
+import (
+	"errors"
+)
+
+type HostResolveImportedModuleFunc func(referencingScriptOrModule interface{}, specifier string) (ModuleRecord, error)
+
+// TODO most things here probably should be unexported and names should be revised before merged in master
+// Record should probably be dropped from everywhere
+
+// ModuleRecord is the common interface for module record as defined in the EcmaScript specification
+type ModuleRecord interface {
+	GetExportedNames(resolveset ...ModuleRecord) []string
+	ResolveExport(exportName string, resolveset ...ResolveSetElement) (*ResolvedBinding, bool)
+	Link() error
+	Evaluate(*Runtime) *Promise
+}
+
+type CyclicModuleRecordStatus uint8
+
+const (
+	Unlinked CyclicModuleRecordStatus = iota
+	Linking
+	Linked
+	Evaluating
+	Evaluating_Async
+	Evaluated
+)
+
+type CyclicModuleRecord interface {
+	ModuleRecord
+	RequestedModules() []string
+	InitializeEnvironment() error
+	Instantiate(rt *Runtime) (CyclicModuleInstance, error)
+}
+
+type (
+	ModuleInstance interface {
+		GetBindingValue(string) Value
+	}
+	CyclicModuleInstance interface {
+		ModuleInstance
+		HasTLA() bool
+		ExecuteModule(rt *Runtime, res, rej func(interface{})) (CyclicModuleInstance, error)
+	}
```

Comment:
The workflow of parsing, compiling to a program and then running on multiple Runtimes has been very useful for k6. We would do these hundreds of times to create a bunch of virtual users (VUs) - which for this use cases are separate goja Runtime that execute code.

With modules, I wanted to keep that. For that I needed to get through with *parsing* and *compilation* and get something that then on execution somehow got all the *compiled* modules together and executed but not change any of the compiled structs.

This took a while to figure out how to do "properly" and is one of the things that I have custom tests for.

I don't really have a question, apart from do you see a different way this could've been done and have the same functionality.